### PR TITLE
Implement #85440 (Random test ordering)

### PR DESCRIPTION
**Pull request by smoelius**
This PR adds `--shuffle` and `--shuffle-seed` options to `libtest`. The options are similar to the [`-shuffle` option](https://github.com/golang/go/blob/c894b442d1e5e150ad33fa3ce13dbfab1c037b3a/src/testing/testing.go#L1482-L1499) that was recently added to Go.

Here are the relevant parts of the help message:
```
        --shuffle       Run tests in random order
        --shuffle-seed SEED
                        Run tests in random order; seed the random number
                        generator with SEED
...
By default, the tests are run in alphabetical order. Use --shuffle or set
RUST_TEST_SHUFFLE to run the tests in random order. Pass the generated
"shuffle seed" to --shuffle-seed (or set RUST_TEST_SHUFFLE_SEED) to run the
tests in the same order again. Note that --shuffle and --shuffle-seed do not
affect whether the tests are run in parallel.
```
Is an RFC needed for this?

### CODE REVIEW

**Comment by smoelius** [comment]
Converting this to a WIP while I figure out this randomness situation.

File: library/test/src/cli.rs [hunk 1]
```diff
@@ -303,6 +322,46 @@ fn get_time_options(
...
+            }
+        },
+        None => None,
+    };
+
+    if shuffle_seed.is_none() {
+        shuffle_seed = match env::var("RUST_TEST_SHUFFLE_SEED") {
+            Ok(val) => match val.parse::<u64>() {
+                Ok(n) => Some(n),
+                Err(_) => panic!("RUST_TEST_SHUFFLE_SEED is `{}`, should be a number.", val),
```
Comment (lukaslueg) [inline library/test/src/cli.rs:356]: Why the `panic!` here instead of returning an Error as above?

Comment (smoelius) [reply library/test/src/cli.rs:356]: I was trying to be consistent with `get_concurrency`: https://github.com/rust-lang/rust/blob/0a0688e6f51229d2860af6317adaef378f4b45a5/library/test/src/helpers/concurrency.rs#L9

But I don't have a strong opinion, and I would be happy to change this.

**Comment by smoelius** [comment]
@kennytm I think you can start reviewing this whenever you are ready.

File: library/test/src/helpers/shuffle.rs [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    fn new(seed: u64, extra: u64) -> Self {
+        Self { state: seed, extra }
+    }
+
+    fn rand_range(&mut self, range: core::ops::Range<u64>) -> u64 {
+        self.rand_u64() % (range.end - range.start) + range.start
+    }
+
+    fn rand_u64(&mut self) -> u64 {
+        self.state = calculate_hash(&(self.state, self.extra));
```
Comment (kennytm) [inline library/test/src/helpers/shuffle.rs:57]: can we use a proper RNG library like `fastrand` or `oorandom` or `rand`

there is no proof that `state = sip13_hash((state, extra))` is a good RNG.

(and additionally `rand_u64() % n` produces  biased result)

File: library/test/src/cli.rs [hunk 3]
```diff
@@ -303,6 +322,46 @@ fn get_time_options(
     Ok(options)
 }
 
+fn get_shuffle(matches: &getopts::Matches) -> OptPartRes<bool> {
+    let mut shuffle = matches.opt_present("shuffle");
```
[the remark this answers is not in the document]
Comment (smoelius) [reply library/test/src/cli.rs]: Sorry, I should have noticed that. Fixed.

File: library/test/src/helpers/shuffle.rs (hunk 2 above)
Comment (smoelius) [reply library/test/src/helpers/shuffle.rs:57]: > can we use a proper RNG library like `fastrand` or `oorandom` or `rand`

I tried `rand` and `oorandom` actually. In both cases, I got `can't find crate for core` errors.

My impression (based on [this](https://github.com/rust-lang/rust/blob/0a0688e6f51229d2860af6317adaef378f4b45a5/library/test/Cargo.toml#L10-L11), [this](https://github.com/alexcrichton/cfg-if/blob/59f5b77e6255ce8baf304f74f141a67d42f339c5/Cargo.toml#L21), and [this](https://github.com/rust-lang/getopts/blob/c11eb65097d0dc9fe9ceb351d63c3ac8b5f05e0b/Cargo.toml#L19)) is that a crate has to go out of its way to be a dependency of a builtin library. And none of those crates seems to do that.

Note that `oorandom` barely uses `core`, but it does use it for `Range` ([here](https://hg.sr.ht/~icefox/oorandom/browse/src/lib.rs?rev=tip#L23)).

> there is no proof that `state = sip13_hash((state, extra))` is a good RNG.
> 
> (and additionally `rand_u64() % n` produces biased result)

Your points are very well taken. However, I don't think this random number generator needs to be "strong." Put another way, I think the risk of an attacker trying to manipulate the test ordering is low. I am open to counterarguments though.

(Thanks for doing this, BTW.)

**Comment by smoelius** [comment]
@kennytm Since we seem to have reached a pause, I took the opportunity to do some squashing.

Please let me know if there is anything I can do to make reviewing this PR easier.

File: library/test/src/cli.rs [hunk 4]
```diff
@@ -303,6 +337,46 @@ fn get_time_options(
     Ok(options)
 }
 
+fn get_shuffle(matches: &getopts::Matches, allow_unstable: bool) -> OptPartRes<bool> {
+    let mut shuffle = unstable_optflag!(matches, allow_unstable, "shuffle");
+    if !shuffle {
+        shuffle = match env::var("RUST_TEST_SHUFFLE") {
```
Comment (kennytm) [inline library/test/src/cli.rs:343]: should the env var be also protected behind `allow_unstable`?

Comment (smoelius) [reply library/test/src/cli.rs:343]: Yes, I agree. Fixed.

**Comment by kennytm** [comment]
please also update [`/src/doc/rustc/src/tests/index.md`](https://github.com/rust-lang/rust/blob/master/src/doc/rustc/src/tests/index.md) to document the new unstable option.

r=me after that.

**Comment by smoelius** [comment]
@kennytm I added some documentation. Please let me know if it seems inadequate.

Thank you again for all of your trouble.

File: src/doc/rustc/src/tests/index.md [hunk 5]
```diff
@@ -181,6 +181,38 @@ unstable-options` flag. See [tracking issue
...
+
+The random number generator seed that is output can be passed to
+[`--shuffle-seed`](#--shuffle-seed-seed) to run the tests in the same order
+again.
+
+Note that `--shuffle` does not affect whether the tests are run in parallel. To
+run the tests in random order sequentially, use `--shuffle --test-threads 1`.
+
+⚠️ 🚧 This option is [unstable](#unstable-options), and requires the `-Z
+unstable-options` flag.
```
Comment (kennytm) [inline src/doc/rustc/src/tests/index.md:199]:
```suggestion
unstable-options` flag. See [tracking issue
#85440](https://github.com/rust-lang/rust/issues/85440) for more 
information.
```

File: src/doc/rustc/src/tests/index.md [hunk 6]
```diff
@@ -181,6 +181,38 @@ unstable-options` flag. See [tracking issue
...
+the tests in the same order both times.
+
+_SEED_ is any 64-bit unsigned integer, for example, one produced by
+[`--shuffle`](#--shuffle).
+
+This can also be specified with the `RUST_TEST_SHUFFLE_SEED` environment
+variable.
+
+⚠️ 🚧 This option is [unstable](#unstable-options), and requires the `-Z
+unstable-options` flag.
```
Comment (kennytm) [inline src/doc/rustc/src/tests/index.md:214]:
```suggestion
unstable-options` flag. See [tracking issue
#85440](https://github.com/rust-lang/rust/issues/85440) for more 
information.
```

**Review by kennytm — APPROVED** [review]
basically I just want a link to a tracking issue. 

or you could file a different issue for tracking if you like 🙃

**Comment by joshtriplett** [comment]
Might be worth looking at lld's   --shuffle-sections for comparison.

**Comment by smoelius** [comment]
> basically I just want a link to a tracking issue.
> 
> or you could file a different issue for tracking if you like upside_down_face

Sorry, I misunderstood how these things work.

I opened a new issue so that I can edit it if I have to.

Thanks again, @kennytm.

**Pull request merged**